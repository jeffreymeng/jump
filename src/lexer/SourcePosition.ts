/**
 * A SourcePosition refers to a specific position in the source.
 * Each character in the source has a unique position.
 */
export default class SourcePosition {
	/**
	 * @param source - The text of the source code.
	 * @param index - The index of the character within the source text.
	 */
	constructor(protected readonly source: string, protected readonly index: number) {
		if (index >= source.length) {
			throw new Error("SourcePosition expected an index less than the length of source.")
		}
	}

	/**
	 * Returns the line of the index in the given source text. Lines are 1-indexed.
	 */
	public getLine(): number {
		return this.source.substring(0, this.index).split("\n").length
	}

	/**
	 * Returns the index of the character within its line. Columns are
	 * 1-indexed.
	 */
	public getCol(): number {
		// given the column, look backwards for the closest newline strictly
		// before the current index, since if the current character is a
		// newline that doesn't count.
		return this.index - this.source.lastIndexOf("\n", this.index - 1);
	}

}
