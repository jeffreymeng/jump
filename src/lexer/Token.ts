export enum TOKEN_TYPE {
	STRING_LITERAL = "string_literal",
	INT_LITERAL = "int_literal",
	DOUBLE_LITERAL = "double_literal",
	OPERATOR = "operator",
	IDENTIFIER = "identifier",
	CONTROL = "control",
	KEYWORD = "keyword",
}
export default class Token {
	/**
	 *
	 * @param type - The type of the token
	 * @param symbol - The value of the token
	 */
	constructor(
		public readonly type: TOKEN_TYPE,
		public readonly symbol: string
	) {}

	/**
	 * Returns whether or not the token matches another token, in terms of content.
	 * @param other
	 */
	public equals(other: Token) {
		return this.type === other.type && this.symbol === other.symbol;
	}

	/**
	 * Returns whether or not the token matches the provided type and symbol.
	 * @param type
	 * @param symbol
	 */
	public is(type: TOKEN_TYPE, symbol: string) {
		return this.type === type && this.symbol === symbol;
	}

	public toString() {
		return `Token<${this.type}, '${this.symbol}'>`;
	}
}
