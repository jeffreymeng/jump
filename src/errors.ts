import SourcePosition from "./lexer/SourcePosition";
import JumpObject from "./ast/objects/JumpObject";

export class JumpError extends Error {
	constructor(message: string) {
		super(message);
	}
}

/**
 * An error in the syntax of the input.
 */
export class JumpSyntaxError extends JumpError {
	public readonly position: SourcePosition;

	constructor(message: string, position: SourcePosition) {
		super(message + "\n\tat " + position.toString());
		this.position = position;
	}
}

// TODO: track position (add position property to each ast node)
export class JumpTypeError extends JumpError {
	// public readonly position: SourcePosition;

	constructor(message: string) {
		super(message);
		// this.position = position;
	}
}

// TODO: track position (add position property to each ast node)
export class JumpNameError extends JumpError {
	// public readonly position: SourcePosition;

	constructor(message: string) {
		super(message);
		// this.position = position;
	}
}

/**
 * Thrown if the provided arguments are invalid in either quantity or
 * value (but not type).
 */
export class JumpArgumentError extends JumpError {
	// public readonly position: SourcePosition;

	constructor(message: string) {
		super(message);
		// this.position = position;
	}
}

export class JumpInternalError extends JumpError {
	constructor(message: string) {
		super(message);
	}
}

export class JumpDivisionByZeroError extends JumpError {
	constructor(dividend: number, public readonly position: SourcePosition) {
		super(`JumpDivisionByZeroError: Cannot divide ${dividend} by zero.`);
	}
}

// not a real error, but we use the error mechanism to get the return value up.
export class JumpReturn extends Error {
	constructor(public readonly returnValue: any = null /*JumpObject (TODO)*/) {
		super("<jump return>");
	}
}
