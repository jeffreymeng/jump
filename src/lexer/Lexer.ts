import Token, { TokenType } from "./Token";
import SourcePosition from "./SourcePosition";
import Source from "./Source";
import { JumpSyntaxError } from "../errors";

export default class Lexer implements IterableIterator<Token> {
	/**
	 * The source code of the lexer.
	 */
	public readonly source: Source;

	/**
	 * The index of the current character within the source.
	 */
	protected index = 0;

	/**
	 * The index of the current token.
	 */
	protected currentTokenIndex = 0;

	constructor(source: Source) {
		this.source = source;
	}

	// TODO: this could be optimized by checking with ascii values.
	public static readonly DIGITS = "0123456789";
	public static readonly LOWERCASE_LETTERS = "abcdefghijklmnopqrstuvwxyz";
	public static readonly UPPERCASE_LETTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

	// TODO: this might be more optimized with a trie
	public static readonly KEYWORDS = ["for", "in", "if", "else", "while"];

	public static readonly OPERATOR_KEYWORDS = ["break", "continue", "return"];

	/**
	 * Returns a new lexer on the same source text but with any internal state reset.
	 */
	public clone(): Lexer {
		return new Lexer(this.source.clone());
	}

	/**
	 * Consumes the next token, and returns it.
	 */
	public nextToken(): Token {
		const source = this.source;
		const c = source.next();
		this.setTokenIndex();
		// TODO: fix issue with comments & done
		// potentially by editing the done function to see if nextToken
		// returns anything (or if c is "" then exit immediately)
		// if (c === "/") {
		// 	// two slashes == comment
		// 	if (source.peekNext() === "/") {
		// 		source.consumeUntil("\n")
		// 	}
		// }

		if ("\"'".includes(c)) {
			// consume until we see the same type of quote again
			const buffer = source.consumeUntil(
				c,
				"Expected string to be closed."
			);
			return this.createToken(TokenType.STRING_LITERAL, buffer);
		}

		if (Lexer.DIGITS.includes(c)) {
			let buffer = c + source.consumeAll(Lexer.DIGITS);
			// if the next character is a dot, then this is part of a double.
			// otherwise, return the int
			if (
				// make sure the next character isn't a dot, or if it is a dot
				// that it's not followed by a number (which would make it not
				// a double).
				source.peekNext() !== "." ||
				// split it so the empty string doesn't trigger this condition
				!Lexer.DIGITS.split("").includes(this.source.peekNext(1))
			) {
				return this.createToken(TokenType.INT_LITERAL, buffer);
			} else {
				buffer += this.source.next() + source.consumeAll(Lexer.DIGITS);
				return this.createToken(TokenType.DOUBLE_LITERAL, buffer);
			}
		}
		if ("+-*/%".includes(c)) {
			let op = c;

			if (source.peekNext() === "*") {
				op = c + source.next();
			}

			// check for math assignment operators (+=, *=, **=, %=, etc.)
			if (source.peekNext() === "=") {
				source.next();
				return this.createToken(TokenType.OPERATOR, op + "=");
			}

			// check for inc/dec (++/--) (these can't be assignment operators)
			if ("+-".includes(c) && source.peekNext() === c) {
				op = c + source.next();
			}

			// return a basic operator
			return this.createToken(TokenType.OPERATOR, op);
		}

		// parentheses can technically be considered both operators and control flow characters.
		// . is an operator outside of number literals.
		if ("[]().<>!=,&|".includes(c)) {
			// == and => (function arrow) operators
			if (c === "=" && "=>".includes(source.peekNext())) {
				return this.createToken(TokenType.OPERATOR, c + source.next());
			}

			// logical operators that can be modified with the equal sign
			if ("<>!".includes(c) && source.peekNext() === "=") {
				return this.createToken(TokenType.OPERATOR, c + source.next());
			}

			// && and || operators
			if ("&|".includes(c) && source.peekNext() === c) {
				return this.createToken(TokenType.OPERATOR, c + source.next());
			}

			return this.createToken(TokenType.OPERATOR, c);
		}
		if ("\n;{}".includes(c)) {
			if (c === "\n") {
				return this.nextToken();
			}
			return this.createToken(TokenType.CONTROL, c);
		}
		if (
			("$_" + Lexer.UPPERCASE_LETTERS + Lexer.LOWERCASE_LETTERS).includes(
				c
			)
		) {
			const buffer =
				c +
				source.consumeAll(
					"$_" +
						Lexer.DIGITS +
						Lexer.UPPERCASE_LETTERS +
						Lexer.LOWERCASE_LETTERS
				);
			if (buffer === "true" || buffer === "false") {
				return this.createToken(TokenType.BOOLEAN_LITERAL, buffer);
			}
			if (Lexer.KEYWORDS.includes(buffer)) {
				return this.createToken(TokenType.KEYWORD, buffer);
			}
			if (Lexer.OPERATOR_KEYWORDS.includes(buffer)) {
				return this.createToken(TokenType.OPERATOR, buffer);
			}
			return this.createToken(TokenType.IDENTIFIER, buffer);
		}
		if (/\s/g.test(c)) {
			// skip any spaces and tabs, and other whitespace (except newlines,
			// which are parsed above).
			return this.nextToken();
		}

		throw new JumpSyntaxError(
			`Unexpected character '${c}'.`,
			source.getPosition()
		);
	}

	public hasNextToken(): boolean {
		// if there are only whitespace characters (including trailing newlines)
		// then we can return.
		return !/^[\s\n]*$/g.test(this.source.peekNextCharacters());
	}

	// TODO: change this to generator *next() while (this.has) return so we don't have to do value stuff ourselves.
	public next(): IteratorResult<Token, null> {
		if (!this.hasNextToken()) {
			return {
				value: null,
				done: true,
			};
		}
		return {
			value: this.nextToken(),
			done: false,
		};
	}

	public getPosition(): SourcePosition {
		return this.source.getPosition();
	}

	/**
	 * Set the current token position to the lexer's next character's position
	 */
	protected setTokenIndex(): void {
		this.currentTokenIndex = this.index;
	}

	public [Symbol.iterator](): IterableIterator<Token> {
		return this;
	}

	protected createToken(type: TokenType, symbol: string) {
		return new Token(type, symbol, this.getPosition());
	}
}
