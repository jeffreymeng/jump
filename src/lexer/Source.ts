import SourcePosition from "./SourcePosition";
import { JumpSyntaxError } from "./Lexer";

export default class Source {
	protected _index = 0;
	public readonly text: string;

	/**
	 * Returns the current index of the source.
	 */
	public get index() {
		return this._index;
	}

	constructor(text: string) {
		this.text = text;
	}

	/**
	 * Returns the next character from the source, updating the line and position as necessary.
	 * If no next character exists, returns an empty string.
	 */
	public next(): string {
		if (this._index >= this.text.length) {
			throw new JumpSyntaxError(
				"Unexpected end of input.",
				new SourcePosition(this.text, this._index)
			);
		}
		const c = this.text[this._index];
		this._index++;
		return c;
	}

	/**
	 * Returns the next character from the source without changing the current position.
	 * @param offset - If provided, the ith character after the current one is returned.
	 * If `i` is out of range, an empty string will be returned.
	 */
	public peekNext(offset = 0): string {
		if (this._index + offset >= this.text.length) {
			return "";
		}

		return this.text[this._index + offset];
	}

	/**
	 * Returns the next `len` characters from the source without changing the current position.
	 * @param len - The number of characters to return, or `0` to return the rest of source. Must be in range
	 */
	public peekNextCharacters(len = 0): string {
		if (len === 0) {
			return this.text.substring(this._index);
		}

		if (this._index + len >= this.text.length) {
			throw new JumpSyntaxError(
				"Unexpected end of file",
				new SourcePosition(this.text, this._index)
			);
		}
		return this.text.substring(this._index, this._index + len);
	}

	/**
	 * Consumes characters until any character inside endChars is found.
	 * The end character is consumed, but only the characters from
	 * the current position (inclusive) to the last character before
	 * the end character are returned. The string must eventually contain
	 * one of the endChars.
	 * @param endChars - A string containing one or more characters that will cause consumeUntil()
	 *  to stop consuming characters.
	 * @param errorMessageIfNotFound - If the string does not eventually contain
	 * an endChar, then this message will be used instead of the default error message.
	 */
	public consumeUntil(
		endChars: string,
		errorMessageIfNotFound = "Unexpected end of input."
	): string {
		const consumed = this.consumeWhile((c) => !endChars.includes(c));
		// consume the character that caused the predicate to fail, unless
		// we reached the end of the string already (and thus nothing caused
		// this to throw), which would cause an error.
		if (this._index < this.text.length) {
			this.next();
		} else {
			throw new JumpSyntaxError(
				errorMessageIfNotFound,
				this.getPosition()
			);
		}
		return consumed;
	}

	/**
	 * Consumes and returns characters until a character not in `chars` is found.
	 * @param chars - A string containing one or more characters to be consumed continuously.
	 */
	public consumeAll(chars: string): string {
		return this.consumeWhile((c) => chars.includes(c));
	}

	/**
	 * Consumes characters until the predicate is no longer true.
	 * All characters from the starting position until the predicate fails will be consumed and returned.
	 * The character that failed the predicate will NOT be consumed.
	 * @param predicate - A function returning true if a character passes the predicate.
	 */
	public consumeWhile(predicate: (c: string) => boolean): string {
		let buffer = "";
		while (this.hasNext() && predicate(this.peekNext())) {
			buffer += this.next();
		}
		return buffer;
	}

	/**
	 * Get a SourcePosition object corresponding to the `i`th character of the source,
	 * or the current index if none is provided.
	 */
	public getPosition(i = this._index) {
		return new SourcePosition(this.text, i);
	}

	/**
	 * Returns true if the source has a next token (e.g. has not reached the end yet).
	 */
	public hasNext() {
		return this._index < this.text.length;
	}
}
