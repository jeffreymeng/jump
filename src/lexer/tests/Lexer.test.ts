import { describe, expect, test } from "@jest/globals";

import Lexer from "../Lexer";
import Source from "../Source";
import Token, { TokenType } from "../Token";
import SourcePosition from "../SourcePosition";
import { JumpSyntaxError } from "../../errors";

describe("Lexer", () => {
	// curried
	const t = (type: TokenType) => (symbol: string) =>
		new Token(type, symbol, new SourcePosition("", 0));
	const int = t(TokenType.INT_LITERAL);
	const str = t(TokenType.STRING_LITERAL);
	const op = t(TokenType.OPERATOR);
	const id = t(TokenType.IDENTIFIER);
	const ctrl = t(TokenType.CONTROL);
	const kw = t(TokenType.KEYWORD);
	const semi = () => ctrl(";");
	const nl = () => ctrl("\n");
	// lexes the string and returns an array of tokens.
	const lexToArray = (s: string): Token[] => [...new Lexer(new Source(s))];
	const lexToArrayMatches = (s: string, expected: Token[]): boolean =>
		lexToArray(s).every((x, i) => x.equals(expected[i]));

	test("Basic Lexing", () => {
		expect(
			lexToArrayMatches(`"hello" 2 0.2 42.294 "23.1"`, [
				str("hello"),
				int("2"),
				t(TokenType.DOUBLE_LITERAL)("0.2"),
				t(TokenType.DOUBLE_LITERAL)("42.294"),
				str("23.1"),
			])
		).toBe(true);
	});

	test("Operators", () => {
		expect(
			lexToArrayMatches(
				`i ++ i *= 5 ** 5 i **= 5 i && x i <= x i || y !=`,
				[
					id("i"),
					op("++"),
					id("i"),
					op("*="),
					int("5"),
					op("**"),
					int("5"),
					id("i"),
					op("**="),
					int("5"),
					id("i"),
					op("&&"),
					id("x"),
					id("i"),
					op("<="),
					id("x"),
					id("i"),
					op("||"),
					id("y"),
					op("!="),
				]
			)
		).toBe(true);
	});

	test("Example", () => {
		const s = `
int[] nums = [1, 2, 3, 4, 5, 9];

for (i in nums.filter(n => n % 2 == 0)) {
	print(i ** 2);
}\n`;

		const expected = [
			nl(),
			id("int"),
			op("["),
			op("]"),
			id("nums"),
			op("="),
			op("["),
			...[1, 2, 3, 4, 5, 9].map((x, i) =>
				i !== 0 ? [op(","), int(x + "")] : [int(x + "")]
			),
			op("]"),
			semi(),
			nl(),
			nl(),
			kw("for"),
			op("("),
			id("i"),
			kw("in"),
			id("nums"),
			op("."),
			id("filter"),
			op("("),
			id("n"),
			op("=>"),
			id("n"),
			op("%"),
			int("2"),
			op("=="),
			int("0"),
			op(")"),
			op(")"),
			ctrl("{"),
			nl(),
			id("print"),
			op("("),
			id("i"),
			op("**"),
			int("2"),
			op(")"),
			semi(),
			nl(),
			ctrl("}"),
			// the lexer does not pass trailing newlines.
		].flat();

		expect(lexToArrayMatches(s, expected)).toBe(true);
	});

	test("Literals", () => {
		expect(lexToArrayMatches(`123.`, [int("123"), op(".")])).toBe(true);
		expect(() => lexToArray(`"hello'`)).toThrowErrorMatchingSnapshot();
		expect(() => lexToArray(`'`)).toThrowErrorMatchingSnapshot();
		expect(lexToArrayMatches(`""'''"'`, [str(""), str(""), str(`"`)])).toBe(
			true
		);
		expect(() => lexToArray(`"'"'`)).toThrowErrorMatchingSnapshot("");
		expect(
			lexToArrayMatches(`8h4i4()`, [
				int("8"),
				id("h4i4"),
				op("("),
				op(")"),
			])
		).toBe(true);
	});

	test("Error cases", () => {
		expect(() => new Lexer(new Source("ðŸ˜Ž")).next()).toThrow(
			JumpSyntaxError
		);
	});

	test("Clone", () => {
		const lex = new Lexer(new Source("1 2 3"));
		expect(lex.nextToken().is(TokenType.INT_LITERAL, "1")).toBe(true);
		expect(lex.nextToken().is(TokenType.INT_LITERAL, "2")).toBe(true);
		expect(lex.clone().nextToken().is(TokenType.INT_LITERAL, "1")).toBe(
			true
		);
	});
});
