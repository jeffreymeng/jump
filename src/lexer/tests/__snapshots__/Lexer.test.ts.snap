// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lexer Literals 1`] = `
"Expected string to be closed.
	at L1:7"
`;

exports[`Lexer Literals 2`] = `
"Expected string to be closed.
	at L1:1"
`;

exports[`Lexer Literals 3`] = `
"Expected string to be closed.
	at L1:4"
`;
